syntax = "proto3";

import "google/protobuf/empty.proto";

package proto.rl_ui_messages;

// ===================================================================
// ============================= Request =============================
// ===================================================================

message SaveCheckpoint {
  string name = 1;
}

message LoadCheckpoint {
  string name = 1;
}

message DeleteCheckpoints {
  repeated string names = 1;
}

message AsyncRequest {
  oneof body {
    google.protobuf.Empty start_training = 1;
    google.protobuf.Empty stop_training = 2;
    google.protobuf.Empty request_checkpoint_list = 3;
    SaveCheckpoint save_checkpoint = 4;
    LoadCheckpoint load_checkpoint = 5;
    DeleteCheckpoints delete_checkpoints = 6;
  }
};

message RequestMessage {
  oneof body {
    google.protobuf.Empty request_broadcast_port = 1;
    google.protobuf.Empty ping = 2;
    AsyncRequest async_request = 3;
  }
}

// ===================================================================
// ============================== Reply ==============================
// ===================================================================

message ReplyMessage {
  oneof body {
    int32 broadcast_port = 1;
    google.protobuf.Empty ping_ack = 2;
    google.protobuf.Empty async_request_ack = 3;
  }
}

// ===================================================================
// ============================ Broadcast ============================
// ===================================================================

message Checkpoint {
  string name = 1;
}

message CheckpointList {
  repeated Checkpoint checkpoints = 1;
}

message PlotData {
  string name = 1;
  float x = 2;
  float y = 3;
}

message CheckpointLoaded {
  string name = 1;
}

message CharacterStatus {
  string name = 1;
  int32 current_hp = 2;
  int32 max_hp = 3;
  int32 current_mp = 4;
  int32 max_mp = 5;
}

message CharacterStatusList {
  repeated CharacterStatus statuses = 1;
}

message BroadcastMessage {
  oneof body {
    google.protobuf.Empty heartbeat = 1;
    CheckpointList checkpoint_list = 2;
    string checkpoint_already_exists = 3;
    google.protobuf.Empty saving_checkpoint = 4;
    PlotData plot_data = 5;
    CheckpointLoaded checkpoint_loaded = 6;
    CharacterStatusList character_status_list = 7;
  }
}

// ===================================================================
// ===================================================================
// ===================================================================